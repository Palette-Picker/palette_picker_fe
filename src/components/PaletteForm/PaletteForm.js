import React, { Component } from 'react';
import { addProject, addPalette } from '../../utils/apiCalls';
import './PaletteForm.scss';

class PaletteForm extends Component {
  constructor() {
    super();
    this.state = {
      colors: [],
      newProjectName: '',
      newPaletteName: '',
      selectedProjectId: null,
      error: ''
    }
  }

  componentDidMount() {
    let colors = [];
    while (colors.length < 5) {
      colors.push(this.getRandomColor())
    }
    colors = colors.map((color, i) => {
      return { [`color${i + 1}`]: color, isLocked: false }
    })
    this.setState({ colors })
  }

  getRandomColor() {
    return "#000000".replace(/0/g,() => {return (~~(Math.random()*16)).toString(16);});
  }

  updateColors = (e) => {
    let { colors } = this.state;
    colors = colors.map((color, i) => {
      if (color.isLocked === false) {
         return {
          [`color${i + 1}`]: this.getRandomColor(), 
          isLocked: false
        }
      } else {
        return color
      }
    })

    this.setState({ colors })
  }

  toggleLock = (index) => {
    const { colors } = this.state;
    const updatedColors = colors.map((color, i) => {
      if (index === i){
        return { 
          [`color${i + 1}`]: color[`color${i + 1}`], 
          isLocked: !color.isLocked
        }
      } else {
        return color;
      }
    })
    this.setState({ colors: updatedColors})
  }

  handleDropDownChange = (e) => {
    // add error handling for not selected

    this.setState({ selectedProjectId: e.target.value});
  }

  handleInputChange = (e) => {
    this.setState({ [e.target.name]: e.target.value })
  }

  handleSubmitProject = async (e) => {
    e.preventDefault();
    const { updateProjects } = this.props;
    const { newProjectName } = this.state;
    try {
      await addProject(newProjectName);
      await updateProjects();
    } catch ({ error }) {
      this.setState({ error })
    };
  }

  handleAddPalette = async (e) => {
    e.preventDefault();
    const { updateProjects } = this.props;
    const { colors, newPaletteName,
      selectedProjectId } = this.state;
    const newPalette = {
      name: newPaletteName,
      color1: colors[0].color1,
      color2: colors[1].color2,
      color3: colors[2].color3,
      color4: colors[3].color4,
      color5: colors[4].color5,
      project_id: selectedProjectId
    };
    try {
      await addPalette(newPalette)
      await updateProjects();
    } catch ({ error }) {
      this.setState({ error })
    };
  }

  clearInputs = () => {

  }

  render() {
    const { projects } = this.props;
    const { colors } = this.state;
    const colorBtns = colors.map((color, i) => {
      const hexCode = color[`color${i + 1}`];
      return <button
            key={hexCode} 
            className='color' 
            style={{backgroundColor: hexCode}}
            onClick={() => this.toggleLock(i)}
            >{hexCode.toUpperCase()} is locked: {color.isLocked.toString()}
        </button>
    });

    const projNames = projects.map(proj => {
      return <option
        key={proj.name}
        value={proj.id}
        >{proj.name}</option>
    });

    return (

      <div>
        <section className='color-section'>
          { colorBtns }
        </section>
        <button 
          className='random'
          onClick={this.updateColors}
        >Randomize!</button>

        <section className='forms'>
          <form>
            <h3>Create a New Project</h3>
            <input 
              className='createProject' 
              name='newProjectName'
              type='text' 
              value={this.newProjectName}
              placeholder='Enter Project Name'
              onChange={this.handleInputChange}
            />
            <button
              onClick={this.handleSubmitProject}
            >Submit</button>
          </form>
          <form>
            <h3>Add this Palette to a Project</h3>
            <select
              value={this.selectedProjectId}
              defaultValue={'default'}
              onChange={this.handleDropDownChange}
            >
            <option value='default' disabled>Choose a Project ...</option>
            { projNames }
            </select>
            <input 
              className='createPalette'
              name='newPaletteName'
              type='text' 
              value={this.newPaletteName}
              placeholder='Enter New Palette Name'
              onChange={this.handleInputChange}
            />
            <button
              onClick={this.handleAddPalette}
            >Add</button>
          </form>
        </section>
      </div>

    )
  }
};

export default PaletteForm;
