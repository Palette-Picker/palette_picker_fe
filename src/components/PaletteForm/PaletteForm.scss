.color-section {
  display: flex;
  justify-content: center;
}

.color {
  height: 300px;
  width: 150px;
  border: 2px black solid;
  border-radius: 5px;
  margin: 10px;
}

.random {
  height: 40px;
  width: 200px;
  margin: 20px;
}

.forms {
  display: flex;
  justify-content: space-around;
}

form {
  width: 40%;
  display: flex;
  flex-direction: column;
  align-items: center;
  button {
    height: 30px;
    width: 50%;
    margin-top: 20px;
  }
}

select {
  height: 30px;
  width: 50%;
}

input {
  height: 30px;
  width: 50%;
}