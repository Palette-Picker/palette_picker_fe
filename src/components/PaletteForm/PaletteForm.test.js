import React from 'react';
import { shallow } from 'enzyme';
import PaletteForm from './PaletteForm';

describe('PaletteForm', () => {

  let wrapper;
  const mockColors = [
    {
      color1: "#82d173",
      isLocked: false
    },
    {
      color2: "#abfaa9",
      isLocked: false
    },
    {
      color3: "#95a3b3",
      isLocked: false
    },
    {
      color4: "#4c2c69",
      isLocked: false
    },
    {
      color5: "#42253b",
      isLocked: false
    }];
  const mockProjects = [
    {
      id: 19,
      name: "Colors",
      palettes: [
        {
          id: 29,
          name: "Purples and greens",
          color1: "#82d173",
          color2: "#abfaa9",
          color3: "#95a3b3",
          color4: "#4c2c69",
          color5: "#42253b",
          project_id: 19,
          created_at: "2019-12-09T00:11:51.362Z",
          updated_at: "2019-12-09T00:11:51.362Z"
        }]
    }];
  const mockPaletteId = null;
  const mockNewPaletteName = '';
  let mockOldProjectName = '';
  const mockSelectedProjectId = null;
  const mockPass = jest.fn();
  const mockUpdateColors = jest.fn();
  const mockToggleLock = jest.fn();
  const mockUpdateProjects = jest.fn();
  const mockEvent = jest.fn();
  const mockPreventEvent = { preventDefault: jest.fn()};

  beforeEach(() => {
    wrapper = shallow(
      <PaletteForm 
        colors={mockColors}
        passPaletteNameAndColors={mockPass}
        updateColors={mockUpdateColors}
        toggleLock={mockToggleLock}
        projects={mockProjects}
        updateProjects={mockUpdateProjects}
        paletteId={mockPaletteId}
        newPaletteName={mockNewPaletteName}
        oldProjectName={mockOldProjectName}
        selectedProjectId={mockSelectedProjectId}
      />)
  });

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should update state when dropdown changes', () => {
    const mockDropdownChange = { target: { name: 'selectedProjectId', value: 19}};
    wrapper.instance().handleDropDownChange(mockDropdownChange);
    expect(wrapper.state('selectedProjectId')).toEqual(19);
  });

  it('should call handleDropDownChange when change happens', () => {
    wrapper.instance().handleDropDownChange = jest.fn();
    wrapper.find('select').simulate('change', mockEvent);
    expect(wrapper.instance().handleDropDownChange).toHaveBeenCalledWith(mockEvent);
  });

  it('should update state when input changes', () => {
    const mockInputChange = { target: { name: 'newProjectName', value: 'A New Project'}};
    wrapper.instance().handleInputChange(mockInputChange);
    expect(wrapper.state('newProjectName')).toEqual('A New Project');
  });

  it('should call handleInputChange when input event occurs', () => {
    const mockInputChange = { target: { name: 'newProjectName', value: 'A New Project'}};
    wrapper.instance().handleInputChange = jest.fn();
    wrapper.find('.create-project').simulate('change', mockInputChange);
    wrapper.instance().forceUpdate();
    expect(wrapper.instance().handleInputChange).toHaveBeenCalled();
  });

  it('should call handleSubmitProject when submit is clicked', () => {
    wrapper.instance().handleSubmitProject = jest.fn();
    wrapper.find('.submit-btn').simulate('click', mockPreventEvent);
    expect(wrapper.instance().handleSubmitProject).toHaveBeenCalledWith(mockPreventEvent);
  });

  it('should call decidePaletteVerb when add button is clicked', () => {
    wrapper.instance().decidePaletteVerb = jest.fn();
    wrapper.find('.add-btn').simulate('click', mockEvent);
    expect(wrapper.instance().decidePaletteVerb).toHaveBeenCalledWith(mockEvent);
  });

  it('should call handleUpdatePalette if oldProjectName has a value after clicking add-btn', () => {
    wrapper.setProps({ oldProjectName: 'Old Name' });
    wrapper.instance().handleUpdatePalette = jest.fn();
    wrapper.instance().decidePaletteVerb(mockPreventEvent);
    expect(wrapper.instance().handleUpdatePalette).toHaveBeenCalledWith(mockPreventEvent)
  });

  it('should call handleAddPalette if oldProjectName is empty string after clicking add-btn', () => {

  });

  it('should call updateColors when random btn is clicked', () => {
    wrapper.find('.random').simulate('click');
    expect(mockUpdateColors).toHaveBeenCalled();
  });

  it('should call toggleLock when a color is clicked', () => {
    const mockIndex = 0;
    wrapper.find('.color').at(0).simulate('click', mockIndex);
    expect(mockToggleLock).toHaveBeenCalledWith(mockIndex);
  });

  it('should set state when clearInputs is called', () => {
    wrapper.setState({ newProjectName: 'New Name'});
    expect(wrapper.state('newProjectName')).toEqual('New Name');
    wrapper.instance().clearInput('newProjectName');
    expect(wrapper.state('newProjectName')).toEqual('');
  });

});