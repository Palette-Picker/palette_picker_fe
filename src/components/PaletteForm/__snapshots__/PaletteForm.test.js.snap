// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaletteForm should match the snapshot 1`] = `
<div>
  <section
    className="color-section"
  >
    <button
      className="color"
      key="#82d173"
      onClick={[Function]}
      style={
        Object {
          "backgroundColor": "#82d173",
        }
      }
    >
      #82D173
       is locked: 
      false
    </button>
    <button
      className="color"
      key="#abfaa9"
      onClick={[Function]}
      style={
        Object {
          "backgroundColor": "#abfaa9",
        }
      }
    >
      #ABFAA9
       is locked: 
      false
    </button>
    <button
      className="color"
      key="#95a3b3"
      onClick={[Function]}
      style={
        Object {
          "backgroundColor": "#95a3b3",
        }
      }
    >
      #95A3B3
       is locked: 
      false
    </button>
    <button
      className="color"
      key="#4c2c69"
      onClick={[Function]}
      style={
        Object {
          "backgroundColor": "#4c2c69",
        }
      }
    >
      #4C2C69
       is locked: 
      false
    </button>
    <button
      className="color"
      key="#42253b"
      onClick={[Function]}
      style={
        Object {
          "backgroundColor": "#42253b",
        }
      }
    >
      #42253B
       is locked: 
      false
    </button>
  </section>
  <button
    className="random"
    onClick={[Function]}
  >
    Randomize!
  </button>
  <section
    className="forms"
  >
    <form>
      <h3>
        Create a New Project
      </h3>
      <input
        className="create-project"
        name="newProjectName"
        onChange={[Function]}
        placeholder="Enter Project Name"
        type="text"
        value=""
      />
      <button
        onClick={[Function]}
      >
        Submit
      </button>
    </form>
    <form>
      <h3>
        Add this Palette to a Project
      </h3>
      <select
        onChange={[Function]}
      >
        <option
          disabled={true}
          selected={true}
          value="default"
        >
          Choose a Project ...
        </option>
        <option
          key="Colors"
          selected={false}
          value={19}
        >
          Colors
        </option>
      </select>
      <input
        className="createPalette"
        name="newPaletteName"
        onChange={[Function]}
        placeholder="Enter New Palette Name"
        type="text"
        value=""
      />
      <button
        className="add-btn"
        onClick={[Function]}
      >
        Add
      </button>
    </form>
  </section>
</div>
`;
